Anchor commitments, new payment basepoint

bitromortac opened the pull request:
This PR implements a newer commitment transaction type [1]. Anchor commitments are commitment transactions that attach a dusty output per realized `to_local`/`to_remote` output, in order for the commitment transaction to be CPFP-bumpable in a force close scenario. Also HTLC transactions are affected, which are fee-amendable allowing for more inputs via the anyonecanpay sighash. The original proposal of anchor commitments [1] was not safe enough [2] which is why `option_anchors_zero_fee_htlc_tx` was devised [3]. Our goal is to go directly to this version, to not having to deal with legacy developments. The PR passes unit and regtest tests and is ready for initial review or feedback (see Todos).

Key changes:
- The `Channel` object now carries a `has_anchors` property which is used to decide on scripts and locktimes, as well as sweeping behavoir.
- Adapt commitment transactions for anchors (anchors, to_remote script 1-CSV, fee estimation, output values) (03-transactions)
- Adapt HTLC transactions (nsequence numbers, 1 CSV-lock, `sighash.ANYONECANPAY+sighash.SINGLE`) (03-transactions)
- New anchor channels have a `payment_basepoint` derived from the funding pubkey and a static secret to be sweepable in the event of a force close due to (onchain) backups

Testing:
- Reproduce signatures from RFC test vectors for commitment and HTLC transactions (requires different sighash flags) (03-transactions)
- Manually tested against lnd v0.11 (last active version of original anchors) - open/close/force close/payment
- New integration test framework to run Electrum peer components inside a single interpreter (better debugging and easier introspection), with additional detailed channel usage scenarios. Tests are not running in sequence right now and are still subject to changes. Some tests are redundant and could be removed.

Other:
- The watchtower is not updated to sweep to_remote as this is not time critical
- Backups don't need an upgrade as we can see at sweep time, whether we deal with a ctx that has anchors

Todo:
- [x] Crosscheck RFC
- [x] Implement zero fee HTLCs and do fee management at sweep time
- [ ] Reserve some UTXOs for fee bumping
- [x] Compatibility testing with zero-fee-htlc (once zero-fee is implemented)
- [x] Wallet upgrading mechanics, compatibilty wallets with old/new channels
- [x] Keep tests for both commitment versions (static_remotekey/anchors)?
- [x] Make commitment type configurable?
- [x] Determine fee for HTLC-transaction bumping
- Sweep the anchor outputs (05-onchain) (omitted for this PR, because we still rely on the `update_fee` message)
- Refactor to have explicit commitment and channel types [4] (another PR) 

[1] Anchors PR: https://github.com/lightningnetwork/lightning-rfc/pull/688
[2] https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-September/002796.html
[3] Zero-fee HLTCs: https://github.com/lightningnetwork/lightning-rfc/pull/824
[4] Channel types: https://github.com/lightningnetwork/lightning-rfc/pull/880

SomberNight commented on electrum/lnsweep.py:
```diff
@@ -328,35 +337,105 @@ def analyze_ctx(chan: 'Channel', ctx: Transaction):
+            ctx: Transaction,
+            sweep_address: str) -> Optional[Dict[str, SweepInfo]]:
+    txs = {}  # type: Dict[str, SweepInfo]
+    """Sweep to_remote, if we use anchors."""
```
Maybe elaborate a bit on the scenario in the docstring.
e.g. 
> If we only have a backup, and the remote force-closed with their ctx, and anchors are enabled, we need to sweep to_remote.

bitromortac replied:
Done

ecdsa commented on electrum/lnchannel.py:
```diff
@@ -525,6 +546,9 @@ def is_static_remotekey_enabled(self) -> bool:
         return False
 
+    def has_anchors(self) -> Optional[bool]:
+        return None
```
is that method used? 
in `create_sweeptx_their_backup_ctx` you are using `lnutil.ctx_has_anchors()` instead

bitromortac replied:
It's not being used, but it's an `abstractmethod`, so has to be implemented. With the `None` return I wanted to indicate that we don't know about the channel having anchors.

SomberNight reviewed (commented):
So many changes :P but looks good overall.
See comments inline.

bitromortac commented on electrum/lnchannel.py:
```diff
@@ -322,9 +324,14 @@ def update_closed_state(self, *, funding_txid: str, funding_height: TxMinedInfo,
+        if self.has_anchors():
+            try:
+                addr = self.lnworker.wallet.get_new_sweep_address_for_channel()
+            except AttributeError:  # fails in tests with MockWallet
```
Done

bitromortac commented on electrum/tests/test_lnutil.py:
```diff
@@ -38,6 +43,285 @@
+    {
+        "Name": "simple commitment tx with no HTLCs",
+        "LocalBalance": 7000000000,
+        "RemoteBalance": 3000000000,
```
Done

bitromortac commented on electrum/lnpeer.py:
```diff
@@ -502,6 +502,10 @@ def is_static_remotekey(self):
 
+    def use_anchors(self) -> bool:
+        return self.features.supports(LnFeatures.OPTION_ANCHOR_OUTPUTS_OPT) \
+               and self.lnworker.features.supports(LnFeatures.OPTION_ANCHOR_OUTPUTS_OPT)
```
Yes, here I wanted to ask how we want to transition to anchors? The second condition was also a reminder for myself.

SomberNight replied:
Well there are three ~reasonable options I can see:
1. we require anchors_zero_fee for newly opened channels
2. we open anchors_zero_fee channels whenever possible, but if the remote does not support it, we can just option static_remotekey channels
3. we disable anchors_zero_fee by default, gate it with a config key; and change to one of the options above in a later release

I think the decision should depend on a number of things, such as
- how many/which nodes on the network will already support anchors_zero_fee,
- how confident we are that our code behaves correctly, 
- how the UX of reserving UTXOs in the wallet looks like,
- whether we can batch HTLC txs (not really for fee-minimisation but rather re number of UTXOs we need to reserve)

So atm I am not sure.

bitromortac commented on electrum/tests/test_lnutil.py:
```diff
@@ -38,6 +43,285 @@
 # funding wscript = 5221023da092f6980e58d2c037173180e9a465476026ee50f96695963e8efe436f54eb21030e9f7b623d2ccc7c9bd44d66d5ce21ce504c0acf6385a132cec6d3c39fa711c152ae
 
 
+# https://github.com/lightningnetwork/lightning-rfc/blob/master/03-transactions.md#appendix-f-commitment-and-htlc-transaction-test-vectors-anchors
```
Agree, changed

bitromortac commented on electrum/lnutil.py:
```diff
@@ -962,6 +1131,18 @@ class LnFeatures(IntFlag):
+    OPTION_ANCHORS_ZERO_FEE_HTLC_OPT = 1 << 23
+    _ln_feature_direct_dependencies[OPTION_ANCHORS_ZERO_FEE_HTLC_OPT] = {OPTION_STATIC_REMOTEKEY_OPT}
+    _ln_feature_contexts[OPTION_ANCHORS_ZERO_FEE_HTLC_REQ] = (LNFC.INIT | LNFC.NODE_ANN)
+    _ln_feature_contexts[OPTION_ANCHORS_ZERO_FEE_HTLC_OPT] = (LNFC.INIT | LNFC.NODE_ANN)
```
Didn't spot that, thanks

bitromortac commented on electrum/lnsweep.py:
```diff
@@ -328,35 +337,105 @@ def analyze_ctx(chan: 'Channel', ctx: Transaction):
+    witness = txin.witness_elements()
+    if len(witness) == 4:
+        # we have OP_2(82) pk1 pk2 OP_2 OP_CHECKMULTISIG(174)
+        # TODO: could match_script_against_template
```
Ok, have done it.

bitromortac commented on electrum/lnsweep.py:
```diff
@@ -328,35 +337,105 @@ def analyze_ctx(chan: 'Channel', ctx: Transaction):
     return ctn, their_pcp, is_revocation, per_commitment_secret
 
 
+def extract_funding_pubkeys_from_ctx(txin: TxInput) -> Optional[List[ecc.ECPubkey]]:
```
Oh yes, initially I returned `ECPubkey`s

bitromortac commented on electrum/lnutil.py:
```diff
@@ -719,41 +798,77 @@ class Direction(IntFlag):
+
+    # if funder cannot afford feerate, their output might go negative, so take max(0, x) here
+    to_local_amt_msat = max(0, to_local_amt_msat)
+    to_remote_amt = max(0, to_remote_amt_msat)
```
Done

bitromortac commented:
I submitted changes via fixup commits, will squash in the end, or should I already?

SomberNight commented:
> I submitted changes via fixup commits, will squash in the end, or should I already?

Feel free to do whatever is convenient.
One remark: if you rebase on master for any reason, it's better to do that in a separate force-push of its own (instead of combining with other changes that would require a force-push). That way, it's easier to see changes in the github UI.

SomberNight commented on electrum/lnsweep.py:
```diff
@@ -328,35 +337,107 @@ def analyze_ctx(chan: 'Channel', ctx: Transaction):
+    We expect to see a witness script of: OP_2 pk1 pk2 OP_2 OP_CHECKMULTISIG"""
+    elements = txin.witness_elements()
+    witness_script = elements[-1]
+    assert match_script_against_template(witness_script, SCRIPTPUBKEY_TEMPLATE_FUNDING)
```
nit. `SCRIPTPUBKEY_TEMPLATE_FUNDING` as a name is a bit confusing:
"scriptPubKey" is conventionally the name for the output script directly visible in the chain (i.e. the script in the output, so what a bitcoin address directly maps to).

So e.g. these are indeed templates for scriptPubKeys:
https://github.com/spesmilo/electrum/blob/437a9e435847431e8654a6e02fc304a714826533/electrum/transaction.py#L436-L442
The templates here are also misnamed IMHO, as they are matching witnessScripts (note difference between witness and witnessScript):
https://github.com/spesmilo/electrum/blob/437a9e435847431e8654a6e02fc304a714826533/electrum/submarine_swaps.py#L37

Maybe just call it `SCRIPT_TEMPLATE_FUNDING`?

bitromortac replied:
Makes sense, done.

bitromortac commented:
> > I submitted changes via fixup commits, will squash in the end, or should I already?
> 
> Feel free to do whatever is convenient. One remark: if you rebase on master for any reason, it's better to do that in a separate force-push of its own just for the rebase (instead of combining with other changes that would require a force-push). That way, it's easier to see changes in the github UI.

I squashed the fixups and rebased on master separately.

There's a new commit, which switches to the `anchors-zero-htlc-fee` type. I have tested this together with the latest LND master. The only issue for now is how to deal with password protected wallets, where we are not able to fee-bump HTLCs automatically (or bump commitment transactions once we phase out `update_fee`).

bitromortac commented:
I reorganized the commits a bit and think this is ready for a second pass. Tests for anchor outputs are disabled right now (can be controlled via `TEST_ANCHOR_CHANNELS`).

SomberNight commented on electrum/lnsweep.py:
```diff
@@ -461,19 +567,97 @@ def create_htlctx_that_spends_from_our_ctx(
+            BIP69_sort=False
+        )
+
+        # sign the fee input(s): TODO: how to deal with password-protected wallets?
```
I see this is not using `wallet.wallet_password` yet

bitromortac replied:
Somehow got lost during rebasing :+1:

bitromortac commented on electrum/gui/qt/installwizard.py:
```diff
@@ -336,7 +337,7 @@ def run_user_interaction_loop():
                                                   _('If you use a passphrase, make sure it is correct.'))
                             self.reset_stack()
-                            return self.select_storage(path, get_wallet_from_daemon)
+                            return
```
Restored.

bitromortac commented on electrum/lnworker.py:
```diff
@@ -648,6 +648,20 @@ def __init__(self, wallet: 'Abstract_Wallet', xprv):
+        # TODO: make this a per-wallet config option?
+        # for anchor commitments we need the password to be able to spend wallet UTXOs
+        if self.config.get('enable_anchor_channels'):
+            if self.wallet.is_watching_only():
```
Have added your suggestions, thank you.

bitromortac commented on electrum/tests/test_lnchannel.py:
```diff
@@ -33,22 +33,23 @@
 
 from . import ElectrumTestCase
 
+TEST_ANCHOR_CHANNELS = False
```
Tests that concern anchor commitment transactions (`test_lnchannel.py`, `test_lnpeer.py`, and `regtest.py`) are now duplicated.

bitromortac commented on electrum/tests/regtest/regtest.sh:
```diff
@@ -2,6 +2,7 @@
 export HOME=~
 set -eu
 
+TEST_ANCHOR_CHANNELS=False
```
Done, I'm setting an environment variable in `subprocess.Popen`. Not sure if updating of the environment dict is the best approach.

bitromortac commented on electrum/tests/integration.py:
```diff
@@ -0,0 +1,555 @@
+import asyncio
```
The tests could be placed into regtest.py and should be executable in the same way (with bitcoind and electrumx requirements). I'm still thinking about which tests to keep. I still have to remove some sleeps and test more for specific outcomes.

bitromortac replied:
I have opened another PR for the tests.

bitromortac commented:
After the recent update all tests passed, but I had to run the watchtower test twice (needs further investigation). I'm still working on the HTLC issue we discussed.

bitromortac commented:
> After the recent update all tests passed, but I had to run the watchtower test twice (needs further investigation). I'm still working on the HTLC issue we discussed.

I tried to add all the edge cases concerning HTLC transaction batching. Do we need an explicit test for this? I also added a test for the watchtower in case of anchor outputs with an HTLC in the ctx (see `integration.TestLightningABC.test_watchtower_htlc`)

ecdsa commented:
Note: I'm going to rebase this in a new branch. Since it is not possible to change the source branch of a PR, I guess this means that the conversation will have to be continued in a new PR

ecdsa commented:
new branch, rebased on current master: https://github.com/spesmilo/electrum/commits/anchor_commitments_2022

ecdsa commented:
new rebase: https://github.com/spesmilo/electrum/commits/anchor_commitments_2024

SomberNight commented:
@ecdsa Could you make a separate PR for your rebase? I cannot usefully add comments to it without that.
(note to self: lnchannel.sweep_address looks broken in the rebase)

ecdsa commented:
opened PR  #9264

ecdsa commented on electrum/lnchannel.py:
```diff
@@ -1193,25 +1193,42 @@ def consider_ctx(*, ctx_owner: HTLCOwner, is_htlc_dust: bool) -> int:
+
+            # initiator pays for anchor outputs
+            if sender == initiator and self.has_anchors():
+                max_send_msat -= 2 * FIXED_ANCHOR_SAT
```
note: needs to be multiplied by 1000

ecdsa commented:
commits were reworked and merged in #9264
